### PR TITLE
Fully implemented authentication with oauth

### PR DESCRIPTION
**Pull request by andy-wojciechowski**
With this PR now if a user tries to navigate to any action in the PostController or an action in a different controller extending BasePostEditorController and they are not authenticated, they will be redirected to GitHub to authorize their GitHub account with the post editor. Once this happens if the user were to navigate the post editor in any future session they will not be required to authorize their account with GitHub again unless their access token has been revoked.

I also changed the index page of the application to be the page which displays the list of posts to edit. The login page has been deleted along with the HomeController.

I haven't tested the authentication with a user that doesn't belong to the msoe-sse organization. In theory, they should be able to at least authorize their account at display the list of posts but they probably will run into errors with editing/adding posts. I wanted to hold off on testing that until we get to implementing that feature.

Once we get this into dev, I think we should merge these changes into master right away because authentication is currently broken on the master branch.

### CODE REVIEW

File: app/controllers/base_post_editor_controller.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  # if we are authenticated by checking to see if a oauth access token is stored in the session. 
+  # If not, we redirect to the GitHub authorization url to authorize a user's GitHub account
+  # with the post editor. If we do have an access token in the session we check to see if it's valid
+  # and if it's not we restart the authenticate flow.
+  def check_authentication
+    if !authenticated?
+      authenticate!
+    else
+      access_token = session[:access_token]
+      is_token_valid = GithubService.check_access_token(access_token)
```
Comment (GFLEMING133) [inline app/controllers/base_post_editor_controller.rb]: Typically variable names are nouns. This one seems to ask a question, so I'd just rename it to 'token_is_valid.'

File: app/controllers/post_controller.rb [hunk 2]
```diff
@@ -1,4 +1,4 @@
-class PostController < ApplicationController
+class PostController < BasePostEditorController
   # GET post/list
   def list
     @posts = GithubService.get_all_posts(session[:access_token])
```
Comment (GFLEMING133) [inline app/controllers/post_controller.rb]: So, question here. This method looks as though it's intended to grab all the current posts from GitHub. In the user stories, we don't technically include a user story for editing existing posts (although I agree it's useful!). That said, I think we can leave the endpoint here for future development. Let's just make sure that the front-end doesn't give this intended functionality away (yet).

File: app/services/github_service.rb [hunk 3]
```diff
@@ -6,28 +6,45 @@
```
Comment (GFLEMING133) [inline app/services/github_service.rb]: Nice work, overall, here.

File: config/routes.rb [hunk 4]
```diff
@@ -1,9 +1,8 @@
 Rails.application.routes.draw do
-  get 'home/index'
-  post 'home/login'
+  get 'github/callback'
   get 'post/list'
   get 'post/edit'
   post 'post/preview'
   post 'post/submit'
-  root 'home#index'
+  root 'post#list'
```
Comment (GFLEMING133) [inline config/routes.rb]: This looks like another instance where it's entirely the fault of the PO (me) for not clarifying exactly what the purpose of the application was. Here it looks like we're grabbing all existing posts to be able to edit those. Is that right?

File: app/controllers/base_post_editor_controller.rb (hunk 1 above)
Comment (andy-wojciechowski) [reply app/controllers/base_post_editor_controller.rb]: Good point, it sounds like I'm setting a method equal to a variable. I renamed the variable to token_valid.

File: app/controllers/post_controller.rb (hunk 2 above)
Comment (andy-wojciechowski) [reply app/controllers/post_controller.rb]: Yes, that controller method currently returning a list of posts to be displayed in post/list. I didn't know that we were not considering editing MVP. In the future I'll be sure that we won't have editing as MVP but I'll keep some of the code that I started for editing.

File: config/routes.rb (hunk 4 above)
Comment (andy-wojciechowski) [reply config/routes.rb]: Yes, I intended for that view to display a list of posts to edit as well as have the functionality to add a new post like through an add button. For MVP I'll probably change the index to the post/edit route

File: app/controllers/post_controller.rb (hunk 2 above)
Comment (GFLEMING133) [reply app/controllers/post_controller.rb]: Awesome, thanks! Again, I'm sorry for the confusion. I guess during your demos, I knew what information you were pulling but I always figured that it was more of an example for showing off the previewing than an actual feature. We'll have to make sure we're on the same page before huge swaths of development in the future.

File: config/routes.rb (hunk 4 above)
Comment (GFLEMING133) [reply config/routes.rb]: That would be perfect.

File: app/controllers/base_post_editor_controller.rb (hunk 1 above)
Comment (GFLEMING133) [reply app/controllers/base_post_editor_controller.rb]: Awesome sauce.

**Comment by andy-wojciechowski** [comment]
@GFLEMING133 The merge conflicts have been resolved related to the rubocop branch being merged into dev. Nothing really major changed in this PR from since you last saw it. I fixed all rubocop offenses in the new authentication code. So, this branch can be merged into dev since I'm done making changes on this branch. 

On a side note did you make sure that the website developers team has write access to this repository? I saw that on your PR review yesterday that your personal GitHub account didn't count for the code review branch and you had to use the webmaster account to approve my PRs.

**Pull request merged**